### PR TITLE
Fix for https://issues.redhat.com/browse/MODCLUSTER-801

### PR DESCRIPTION
**Pull request by jfclere**
Fix for https://issues.redhat.com/browse/MODCLUSTER-801

mod_proxy_cluster returns 404 now instead of 503 when there's no matching context

No upstream commit as return 503 is not correct in the case!

### CODE REVIEW

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 1]
```diff
@@ -3617,12 +3618,18 @@ static int proxy_cluster_pre_request(proxy_worker **worker,
 
                 return HTTP_SERVICE_UNAVAILABLE;
             } else {
-                ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, r->server,
+                int loglevel = APLOG_DEBUG;
+                 apr_status_t error_code = HTTP_NOT_FOUND;
+                if (return503_when_no_context) {
+                    loglevel = APLOG_ERR;
+                    error_code = HTTP_SERVICE_UNAVAILABLE;
+                }
```
Comment (jajik) [inline native/mod_proxy_cluster/mod_proxy_cluster.c:3626]: The indentation is not ok here (additional space).

I would personally write
```c
int log_level = return503_when_no_context ? APLOG_ERR : APLOG_DEBUG;
error_code = return503_when_no_context ? HTTP_SERVICE_UNAVAILABLE : HTTP_NOT_FOUND;
```

but that's a matter of preference.

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 2]
```diff
@@ -4007,6 +4020,14 @@ static const command_rec  proxy_cluster_cmds[] =
         OR_ALL,
         "UseNocanon - When no ProxyPass or ProxyMatch for the URL, passes the URL path \"raw\" to the backend (Default: Off)"
     ),
+
+    AP_INIT_FLAG(
+        "Return503WhenNoContext",
```
Comment (jajik) [inline native/mod_proxy_cluster/mod_proxy_cluster.c:4025]: Maybe `"NoContext503"` would be less wordy?

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 3]
```diff
@@ -4007,6 +4020,14 @@ static const command_rec  proxy_cluster_cmds[] =
...
+        cmd_proxy_cluster_return503_when_no_context,
+        NULL,
+        OR_ALL,
+        "Return503WhenNoContext - When ProxyPass or ProxyMatch doesn't have matching context return 503 (Default: Off)"
```
Comment (jajik) [inline native/mod_proxy_cluster/mod_proxy_cluster.c:4029]: Maybe reword this to

> Return 503 when ProxyPass or ProxyMatch don't have matching context (Default: Off)

**Review by jajik — COMMENTED** [review]
Few details, but ok overall.

Do we want to support this in 2.0 too? If so, I would consider the proposed naming.

File: native/mod_proxy_cluster/mod_proxy_cluster.c (hunk 2 above)
Comment (rhusar) [reply native/mod_proxy_cluster/mod_proxy_cluster.c:4025]: +1 I think we need better name for this as 'Return503WhenNoContext' is very cumbersome.

**Review by rhusar — CHANGES_REQUESTED** [review]

**Comment by jfclere** [comment]
Return503WhenNoContext.... Return503NoContext?

**Comment by jajik** [comment]
`NoContext503`, `NoContextCode503`, `NoContextGive503`, `NoContextResp503`...

In theory, this could be parametrized, require one argument – the value, then it could be something like `NoContextResponse`, `NoContextStatus`, `NoContextCode`... + an argument.

**Comment by rhusar** [comment]
> `NoContext503`, `NoContextCode503`, `NoContextGive503`, `NoContextResp503`...
> 
> In theory, this could be parametrized, require one argument – the value, then it could be something like `NoContextResponse`, `NoContextStatus`, `NoContextCode`... + an argument.

Then we can call it something like - 'ResponseStatusCodeOnNoContext' / 'ResponseStatusCodeWhenNoContext'. The question though is which of the codes are appropriate in this case anyway? Looking at e.g. https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_server_errors, perhaps 502 or 504 - if there were a timeout waiting for contexts to register - which there should be anyway, no?

**Comment by rhusar** [comment]
@jfclere Shouldn't we be attempting to wait until context is available? IIUC we don't but we should - we are expecting to servers boot again and register again.

**Comment by jfclere** [comment]
> @jfclere Shouldn't we be attempting to wait until context is available? IIUC we don't but we should - we are expecting to servers boot again and register again.

No we might wait "for ever" and get deny of service too easily.

**Comment by sgala** [comment]
> > `NoContext503`, `NoContextCode503`, `NoContextGive503`, `NoContextResp503`...
> > In theory, this could be parametrized, require one argument – the value, then it could be something like `NoContextResponse`, `NoContextStatus`, `NoContextCode`... + an argument.
> 
> Then we can call it something like - 'ResponseStatusCodeOnNoContext' / 'ResponseStatusCodeWhenNoContext'. The 

I like the idea, it is clear on what the configuration does

> question though is which of the codes are appropriate in this case anyway other than 503? Looking at e.g. https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_server_errors, perhaps 502 or 504 - if there were a timeout waiting for contexts to register - which there should be anyway, no?

502 seems assumes that an invalid response was received from the server proxied, so I guess it should be 503 or 504.

**Review by jajik — CHANGES_REQUESTED** [review]
LGTM overall, but wouldn't it be better to have

```c
static int responsecode_when_no_context = HTTPD_NOT_FOUND;
```

and then the block could be replaced with single

```c
ap_log_error(APLOG_MARK, response_when_no_context == HTTPD_NOT_FOUND ? APLOG_DEBUG : 
             "proxy: CLUSTER: (%s). No context for the URL returns %d",
             (*balancer)->s->name, response_when_no_context);
```

?

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 4]
```diff
@@ -3617,12 +3618,18 @@ static int proxy_cluster_pre_request(proxy_worker **worker,
...
                 return HTTP_SERVICE_UNAVAILABLE;
             } else {
-                ap_log_error(APLOG_MARK, APLOG_DEBUG, 0, r->server,
+                int loglevel = APLOG_DEBUG;
+                 apr_status_t error_code = HTTP_NOT_FOUND;
+                if (responsecode_when_no_context) {
+                    loglevel = APLOG_ERR;
+                    error_code = responsecode_when_no_context;
+                }
+                ap_log_error(APLOG_MARK, error_code, 0, r->server,
```
[the remark this answers is not in the document]
Comment (jfclere) [reply native/mod_proxy_cluster/mod_proxy_cluster.c]: fixed.

File: native/mod_proxy_cluster/mod_proxy_cluster.c [hunk 5]
```diff
@@ -3947,6 +3948,18 @@ static const char *cmd_proxy_cluster_use_nocanon(cmd_parms *parms, void *mconfig
...
+static const char *cmd_proxy_cluster_responsecode_when_no_context(cmd_parms *parms, void *mconfig, const char *arg)
+{
+    int val = atoi(arg);
+    if (val<0) {
+        return "ResponseStatusCodeOnNoContext must be greater than 0";
+    } else {
+        responsecode_when_no_context = val;
+    }
+
+
```
Comment (jajik) [inline native/mod_proxy_cluster/mod_proxy_cluster.c:3960]: additional blank line

**Review by jajik — APPROVED** [review]
LGTM. Could you please just remove the unnecessary double blank space and squash those four commits into one before merging?

**Comment by rhusar** [comment]
As @jajik says, please squash into a single commit.

**Comment by rhusar** [comment]
@jfclere Can you please prepare PR for https://github.com/modcluster/mod_proxy_cluster ?

**Comment by rhusar** [comment]
@aogburn Could you please give this a quick look (given you are the original reporter)? Thanks!
